File: packages/pylibs/dip/src/dip/framing.py
"""The DIP wire framing: a prologue, a chunked control block, a chunked payload.

Transport is ``AF_UNIX`` / ``SOCK_SEQPACKET``: the kernel preserves message boundaries and
ordering, so a message needs no length prefix -- only a count of the bytes that follow it.

A message is a JSON control block plus an optional binary payload (the image bytes on
``infer``). **Both are chunked**, because a single AF_UNIX datagram cannot exceed
``SO_SNDBUF`` (212992 bytes on a default Linux kernel) -- ``send`` fails with EMSGSIZE
above that, it does not fragment. A dense page yields thousands of lines, so the control
block hits that ceiling as readily as an image does.

    datagram 0     : the prologue -- a small fixed-shape JSON object, always well under
                     the ceiling: {"protocol", "control_len", "payload_len"}
    next datagrams : the control block, in chunks of at most MAX_CHUNK bytes
    next datagrams : the payload, in chunks of at most MAX_CHUNK bytes

A length of 0 means that section sends no datagrams at all. Since no conforming datagram
is ever empty, an empty read means the peer closed.

MAX_CHUNK is the only size a peer has to agree on, and ``handshake`` advertises it along
with the control and payload ceilings. Every function here takes the ``Limits`` in force,
because a requester frames with what its peer advertised rather than with these defaults.
The receive buffer is exactly that ``max_chunk``; a peer that sends a larger datagram is
caught by MSG_TRUNC rather than silently truncated.

Decoding is written against a datagram reader rather than a socket, so the conformance
corpus can drive the same code the socket drives. ``socket_reader`` is the only place that
knows about a file descriptor.
"""

from __future__ import annotations

import json
import socket
from collections.abc import Callable, Mapping
from dataclasses import dataclass, replace
from typing import Any

PROTOCOL_VERSION = 2

# Comfortably under the default SO_SNDBUF so a chunk always fits in one datagram. This is
# the default only: a connection frames with the `Limits` it negotiated, and the receive
# buffer is that connection's `max_chunk`.
MAX_CHUNK = 64 * 1024
RECV_BUFFER = MAX_CHUNK

# Ceilings, mostly to keep a malformed peer from making us allocate forever. 8 MiB of
# control block is roughly 70k OCR lines.
MAX_CONTROL = 8 * 1024 * 1024
MAX_PAYLOAD = 64 * 1024 * 1024
MAX_PROLOGUE = 4096

# Seconds. IDLE applies while waiting for the next message on an open connection;
# MESSAGE applies once a prologue has been read and the rest of the message is owed.
IDLE_TIMEOUT = 300.0
MESSAGE_TIMEOUT = 30.0
SEND_TIMEOUT = 30.0

# One datagram as the kernel hands it over: the bytes, and whether more arrived than the
# receive buffer could hold. A reader raises Timeout when the peer goes quiet and returns
# empty bytes when it closes.
DatagramReader = Callable[[float | None], tuple[bytes, bool]]


class ProtocolError(Exception):
    """The peer sent something that is not a valid message."""


class Timeout(Exception):
    """The peer went quiet in the middle of a message, or never sent one."""


class PeerGone(Exception):
    """The peer closed the connection."""


@dataclass(frozen=True)
class Limits:
    """The sizes a connection frames with. Nothing on the wire may be hard-coded: a
    receiver advertises its own in `handshake`, and a requester adopts what it is told."""

    max_chunk: int = MAX_CHUNK
    max_control: int = MAX_CONTROL
    max_payload: int = MAX_PAYLOAD
    idle_timeout_s: int = int(IDLE_TIMEOUT)
    message_timeout_s: int = int(MESSAGE_TIMEOUT)

    def as_dict(self) -> dict[str, int]:
        return {
            "max_chunk": self.max_chunk,
            "max_control": self.max_control,
            "max_payload": self.max_payload,
            "idle_timeout_s": self.idle_timeout_s,
            "message_timeout_s": self.message_timeout_s,
        }

    def adopt(self, advertised: Mapping[str, Any]) -> "Limits":
        """These limits with whatever the peer advertised laid over them. A field it left
        out, or left at zero, keeps the current value: a missing limit is not a limit of
        nothing."""
        known = self.as_dict()
        taken = {
            field: value
            for field, value in advertised.items()
            if field in known and isinstance(value, int) and not isinstance(value, bool) and value > 0
        }
        return replace(self, **taken)


DEFAULT_LIMITS = Limits()


def limits() -> dict[str, int]:
    """What a peer needs to know to talk to us. Advertised by `handshake`."""
    return DEFAULT_LIMITS.as_dict()


def encode_message(
    control: dict[str, Any], payload: bytes = b"", limits: Limits = DEFAULT_LIMITS
) -> list[bytes]:
    """The datagrams one message becomes, prologue first. Over a ceiling is a ProtocolError."""
    control_blob = json.dumps(control, ensure_ascii=False).encode("utf-8")
    if len(control_blob) > limits.max_control:
        raise ProtocolError(
            f"control block is {len(control_blob)} bytes, over the {limits.max_control} limit"
        )
    if len(payload) > limits.max_payload:
        raise ProtocolError(f"payload is {len(payload)} bytes, over the {limits.max_payload} limit")

    prologue = json.dumps(
        {
            "protocol": PROTOCOL_VERSION,
            "control_len": len(control_blob),
            "payload_len": len(payload),
        }
    ).encode("utf-8")
    chunk = limits.max_chunk
    return [prologue, *_chunks(control_blob, chunk), *_chunks(payload, chunk)]


def send_message(
    sock: socket.socket,
    control: dict[str, Any],
    payload: bytes = b"",
    timeout: float | None = SEND_TIMEOUT,
    limits: Limits = DEFAULT_LIMITS,
) -> None:
    datagrams = encode_message(control, payload, limits)

    previous = sock.gettimeout()
    sock.settimeout(timeout)
    try:
        for datagram in datagrams:
            sock.send(datagram)
    except TimeoutError as exc:
        raise Timeout("peer stopped reading") from exc
    finally:
        sock.settimeout(previous)


def recv_message(
    sock: socket.socket,
    idle_timeout: float | None = IDLE_TIMEOUT,
    message_timeout: float | None = MESSAGE_TIMEOUT,
    limits: Limits = DEFAULT_LIMITS,
) -> tuple[dict[str, Any], bytes]:
    return read_message(socket_reader(sock, limits), idle_timeout, message_timeout, limits)


def socket_reader(sock: socket.socket, limits: Limits = DEFAULT_LIMITS) -> DatagramReader:
    """Bind the framing to a real socket. MSG_TRUNC is the kernel's truncation flag."""

    def read(timeout: float | None) -> tuple[bytes, bool]:
        previous = sock.gettimeout()
        sock.settimeout(timeout)
        try:
            data, _ancillary, flags, _address = sock.recvmsg(limits.max_chunk)
        except TimeoutError as exc:
            raise Timeout(f"peer sent nothing for {timeout}s") from exc
        finally:
            sock.settimeout(previous)
        return data, bool(flags & socket.MSG_TRUNC)

    return read


def read_message(
    read: DatagramReader,
    idle_timeout: float | None = IDLE_TIMEOUT,
    message_timeout: float | None = MESSAGE_TIMEOUT,
    limits: Limits = DEFAULT_LIMITS,
) -> tuple[dict[str, Any], bytes]:
    raw = _read_datagram(read, idle_timeout, limits)
    if not raw:
        raise PeerGone("peer closed the connection")
    if len(raw) > MAX_PROLOGUE:
        raise ProtocolError(f"prologue is {len(raw)} bytes; expected a small JSON header")

    prologue = _decode_json_object(raw, "prologue")
    _check_version(prologue)
    control_len = _length(prologue, "control_len", limits.max_control)
    payload_len = _length(prologue, "payload_len", limits.max_payload)
    if control_len == 0:
        raise ProtocolError("prologue announced an empty control block")

    control_blob = _read_exact(read, control_len, message_timeout, limits)
    payload = _read_exact(read, payload_len, message_timeout, limits)
    return _decode_json_object(control_blob, "control block"), payload


def _chunks(blob: bytes, max_chunk: int) -> list[bytes]:
    return [blob[start : start + max_chunk] for start in range(0, len(blob), max_chunk)]


def _read_datagram(read: DatagramReader, timeout: float | None, limits: Limits) -> bytes:
    data, truncated = read(timeout)
    if truncated:
        raise ProtocolError(
            f"peer sent a datagram larger than the {limits.max_chunk} byte chunk limit"
        )
    return data


def _read_exact(read: DatagramReader, total: int, timeout: float | None, limits: Limits) -> bytes:
    if total == 0:
        return b""

    chunks = []
    received = 0
    while received < total:
        chunk = _read_datagram(read, timeout, limits)
        if not chunk:
            raise PeerGone(f"peer closed after {received} of {total} announced bytes")
        chunks.append(chunk)
        received += len(chunk)
    if received != total:
        raise ProtocolError(f"peer sent {received} bytes, prologue announced {total}")
    return b"".join(chunks)


def _decode_json_object(raw: bytes, what: str) -> dict[str, Any]:
    try:
        decoded = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProtocolError(f"{what} is not valid JSON: {exc}") from exc
    if not isinstance(decoded, dict):
        raise ProtocolError(f"{what} must be a JSON object, got {type(decoded).__name__}")
    return decoded


def _check_version(prologue: dict[str, Any]) -> None:
    """A version the peer does not announce is this one -- the field is optional, and the
    corpus says so. A version it does announce and we do not speak is the single failure
    the field exists to catch, so it is refused rather than served."""
    version = prologue.get("protocol")
    if version is not None and version != PROTOCOL_VERSION:
        raise ProtocolError(
            f"peer speaks protocol {version!r}, this package speaks {PROTOCOL_VERSION}"
        )


def _length(prologue: dict[str, Any], field: str, ceiling: int) -> int:
    if field not in prologue:
        # Defaulting to zero would turn a peer that forgot the payload into a peer that
        # announced an empty one, and the message after it would be read as this one's.
        raise ProtocolError(f"prologue is missing {field}, which every message must announce")
    value = prologue[field]
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise ProtocolError(f"prologue {field} must be a non-negative integer")
    if value > ceiling:
        raise ProtocolError(f"prologue {field} of {value} exceeds the {ceiling} byte limit")
    return value

File: packages/golibs/dip/framing.go
// Package dip implements the Dita Inference Protocol, version dip/2: a request/response
// protocol for inference workers on the same host, carried over AF_UNIX / SOCK_SEQPACKET.
//
// A message is a JSON control block plus an optional opaque payload, framed as:
//
//	datagram 0     : the prologue -- {"protocol": 2, "control_len": N, "payload_len": M}
//	next datagrams : the control block, in chunks of at most max_chunk bytes
//	next datagrams : the payload, in chunks of at most max_chunk bytes
//
// A length of zero means that section sends no datagrams at all, and no conforming
// datagram is ever empty, so an empty read means the peer closed.
//
// Both sections are chunked because a single AF_UNIX datagram cannot exceed SO_SNDBUF
// (212992 bytes on a default Linux kernel) -- send fails with EMSGSIZE above that, it does
// not fragment. A dense page's OCR result runs past that ceiling as readily as an image
// does, so chunking only the payload works until the first busy page.
//
// The prose specification is docs/protocol/[1]dip-specification.md, the IDL is
// specs/dip/dip.schema.json, and the corpus in specs/dip/conformance/ is what keeps this
// implementation and the Python one agreeing. types.go is generated from the IDL.
//
// Roles: Requester is implemented here. The receiver is documented in receiver.go and is
// not implemented, because nothing needs it yet.
package dip

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"slices"
)

// ProtocolVersion is the wire version this package speaks. A peer answering a different
// version is a peer this package should not drive.
const ProtocolVersion = 2

// maxPrologue bounds the first datagram of a message. The prologue is small and
// fixed-shape by design, so anything larger is a peer that has misunderstood the framing
// and is refused before its lengths are trusted.
const maxPrologue = 4096

// DefaultLimits are the limits to frame with before handshake has answered. They are the
// values this implementation's receivers advertise; a requester replaces them with
// whatever its peer reports, because nothing on the wire may be hard-coded.
func DefaultLimits() Limits {
	return Limits{
		MaxChunk:        64 * 1024,
		MaxControl:      8 * 1024 * 1024,
		MaxPayload:      64 * 1024 * 1024,
		IdleTimeoutS:    300,
		MessageTimeoutS: 30,
	}
}

// Error is a failure a peer named: the stable code and the prose that accompanies it.
// Branch on Code; Message is for humans and logs.
type Error struct {
	Code    ErrorCode
	Message string
}

// Error implements the error interface.
func (e *Error) Error() string { return string(e.Code) + ": " + e.Message }

// CodeOf reports the DIP error code carried by err, or the empty string when err is not a
// protocol failure. It is the branch point for a caller deciding whether to retry: a
// checksum_mismatch or a fetch_failed will not be fixed by repeating the infer.
func CodeOf(err error) ErrorCode {
	var protocolErr *Error
	if errors.As(err, &protocolErr) {
		return protocolErr.Code
	}
	return ""
}

// ErrIncomplete reports that the datagrams ended in the middle of a message: the peer
// closed or went quiet with bytes still owed. It is not a refusal, so it carries no error
// code -- there was never a complete message to refuse.
var ErrIncomplete = errors.New("message incomplete")

func badRequest(format string, args ...any) *Error {
	return &Error{Code: ErrorCodeBadRequest, Message: fmt.Sprintf(format, args...)}
}

// A datagramSource yields exactly one whole datagram per call, in order. It exists so the
// framing can be driven by the conformance corpus as readily as by a socket.
type datagramSource interface {
	next() ([]byte, error)
}

// A datagramSink writes exactly one datagram per call. One Write is one datagram on
// SOCK_SEQPACKET, which is the whole reason this is not an io.Writer: an io.Writer may
// coalesce, and coalescing destroys the framing.
type datagramSink interface {
	send(datagram []byte) error
}

// packetConn carries DIP messages over a SOCK_SEQPACKET connection.
type packetConn struct {
	conn *net.UnixConn
	// buf is one byte longer than the chunk limit. SOCK_SEQPACKET discards the tail of a
	// datagram that does not fit, so a full buffer is how an over-limit datagram is
	// caught here -- the equivalent of the MSG_TRUNC flag the C API reports.
	buf []byte
}

func newPacketConn(conn *net.UnixConn, limits Limits) *packetConn {
	return &packetConn{conn: conn, buf: make([]byte, limits.MaxChunk+1)}
}

// resize adopts a peer's chunk limit, which arrives with the handshake.
func (p *packetConn) resize(limits Limits) {
	if want := limits.MaxChunk + 1; want > 0 && want != len(p.buf) {
		p.buf = make([]byte, want)
	}
}

func (p *packetConn) next() ([]byte, error) {
	n, err := p.conn.Read(p.buf)
	if err != nil {
		if errors.Is(err, io.EOF) {
			return nil, fmt.Errorf("%w: peer closed the connection", ErrIncomplete)
		}
		var netErr net.Error
		if errors.As(err, &netErr) && netErr.Timeout() {
			return nil, fmt.Errorf("%w: %w", ErrIncomplete, err)
		}
		return nil, err
	}
	if n == 0 {
		// No conforming datagram is empty, so this is the peer closing.
		return nil, fmt.Errorf("%w: peer closed the connection", ErrIncomplete)
	}
	out := make([]byte, n)
	copy(out, p.buf[:n])
	return out, nil
}

func (p *packetConn) send(datagram []byte) error {
	_, err := p.conn.Write(datagram)
	return err
}

// writeMessage sends one message: the prologue in a single datagram, then the control
// block and the payload, each split into datagrams of at most limits.MaxChunk bytes.
func writeMessage(dst datagramSink, control any, payload []byte, limits Limits) error {
	body, err := json.Marshal(control)
	if err != nil {
		return fmt.Errorf("encode control block: %w", err)
	}
	if len(body) > limits.MaxControl {
		return fmt.Errorf("control block is %d bytes, over the peer's %d byte limit", len(body), limits.MaxControl)
	}
	if len(payload) > limits.MaxPayload {
		return fmt.Errorf("payload is %d bytes, over the peer's %d byte limit", len(payload), limits.MaxPayload)
	}

	version := ProtocolVersion
	head, err := json.Marshal(Prologue{
		Protocol:   &version,
		ControlLen: len(body),
		PayloadLen: len(payload),
	})
	if err != nil {
		return fmt.Errorf("encode prologue: %w", err)
	}

	if err := dst.send(head); err != nil {
		return fmt.Errorf("write prologue: %w", err)
	}
	if err := sendChunked(dst, body, limits.MaxChunk); err != nil {
		return fmt.Errorf("write control block: %w", err)
	}
	if err := sendChunked(dst, payload, limits.MaxChunk); err != nil {
		return fmt.Errorf("write payload: %w", err)
	}
	return nil
}

func sendChunked(dst datagramSink, blob []byte, maxChunk int) error {
	for start := 0; start < len(blob); start += maxChunk {
		end := min(start+maxChunk, len(blob))
		if err := dst.send(blob[start:end]); err != nil {
			return err
		}
	}
	return nil
}

// readMessage reads one message, reassembling both chunked sections. It returns the raw
// control block -- validated as a JSON object, not decoded into any particular op -- and
// the payload.
//
// Three outcomes, which the conformance corpus names: a message, an *Error refusing what
// the peer sent, or an error wrapping ErrIncomplete when the datagrams end mid-message.
func readMessage(src datagramSource, limits Limits) (json.RawMessage, []byte, error) {
	raw, err := src.next()
	if err != nil {
		return nil, nil, fmt.Errorf("read prologue: %w", err)
	}
	if len(raw) > limits.MaxChunk {
		return nil, nil, badRequest("peer sent a datagram over the %d byte chunk limit", limits.MaxChunk)
	}
	if len(raw) > maxPrologue {
		return nil, nil, badRequest("prologue is %d bytes; expected a small JSON header", len(raw))
	}

	// The generated Prologue carries the IDL's own rules: both lengths required, integral
	// and non-negative, and a control block of at least one byte. The ceilings below are
	// framing policy rather than shape, so they live here.
	var head Prologue
	if err := json.Unmarshal(raw, &head); err != nil {
		return nil, nil, badRequest("prologue is not a dip/%d header: %v", ProtocolVersion, err)
	}
	if head.Protocol != nil && *head.Protocol != ProtocolVersion {
		return nil, nil, badRequest("peer speaks protocol %d, this package speaks %d", *head.Protocol, ProtocolVersion)
	}
	if head.ControlLen > limits.MaxControl {
		return nil, nil, badRequest("announced control block of %d bytes exceeds the %d byte limit", head.ControlLen, limits.MaxControl)
	}
	if head.PayloadLen > limits.MaxPayload {
		return nil, nil, badRequest("announced payload of %d bytes exceeds the %d byte limit", head.PayloadLen, limits.MaxPayload)
	}

	control, err := readExactly(src, head.ControlLen, limits)
	if err != nil {
		return nil, nil, fmt.Errorf("read control block: %w", err)
	}
	payload, err := readExactly(src, head.PayloadLen, limits)
	if err != nil {
		return nil, nil, fmt.Errorf("read payload: %w", err)
	}

	var object map[string]json.RawMessage
	if err := json.Unmarshal(control, &object); err != nil {
		return nil, nil, badRequest("control block is not valid JSON: %v", err)
	}
	if object == nil {
		return nil, nil, badRequest("control block must be a JSON object, got null")
	}
	return control, payload, nil
}

// readExactly reassembles total bytes from whole datagrams. The announced length sizes
// nothing: a peer that announces 64 MB and sends nothing should cost this side nothing.
func readExactly(src datagramSource, total int, limits Limits) ([]byte, error) {
	if total == 0 {
		return nil, nil
	}
	out := make([]byte, 0, min(total, limits.MaxChunk))
	for len(out) < total {
		chunk, err := src.next()
		if err != nil {
			return nil, err
		}
		if len(chunk) > limits.MaxChunk {
			return nil, badRequest("peer sent a datagram over the %d byte chunk limit", limits.MaxChunk)
		}
		out = append(out, chunk...)
	}
	if len(out) != total {
		return nil, badRequest("peer sent %d bytes, prologue announced %d", len(out), total)
	}
	return out, nil
}

// Op is a DIP operation: the one field every control block carries.
type Op string

// The ops this protocol version defines. handshake and version are the same op under two
// names, and the health probes are ops rather than URL paths because there is no HTTP
// surface to hang paths on.
const (
	OpHandshake Op = "handshake"
	OpVersion   Op = "version"
	OpList      Op = "list"
	OpLoad      Op = "load"
	OpUnload    Op = "unload"
	OpInfer     Op = "infer"
	OpLivez     Op = "livez"
	OpReadyz    Op = "readyz"
	OpStartupz  Op = "startupz"
)

// opFields is what each op declares beyond "op" itself. Anything undeclared is refused
// rather than ignored: silent tolerance hides typos and turns a client bug into a server
// behaviour. The table mirrors op_fields in specs/dip/conformance/dispatch.json.
var opFields = map[Op][]string{
	OpHandshake: nil,
	OpVersion:   nil,
	OpList:      nil,
	OpLoad:      {"id", "model"},
	OpUnload:    nil,
	OpInfer:     nil,
	OpLivez:     nil,
	OpReadyz:    nil,
	OpStartupz:  nil,
}

// ValidateControl applies the dispatch rule to a decoded control block: the op must exist
// and every other field must be one the op declares. It returns the op, or an *Error with
// code bad_request naming what was wrong.
//
// This is the receiver's judgement, and it lives here rather than in receiver.go because
// it is protocol rather than role -- the conformance corpus asserts it against both
// implementations, and the Go receiver will call this when it lands. A requester's typed
// ops are valid by construction, so Requester.Call does not gate on it: the receiver is
// the authority on dispatch, and a receiver may add an op without a version bump.
func ValidateControl(control json.RawMessage) (Op, error) {
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(control, &fields); err != nil {
		return "", badRequest("control block is not valid JSON: %v", err)
	}
	if fields == nil {
		return "", badRequest("control block must be a JSON object")
	}

	rawOp, ok := fields["op"]
	if !ok {
		return "", badRequest("control block has no op field")
	}
	var name string
	if err := json.Unmarshal(rawOp, &name); err != nil {
		return "", badRequest("op must be a string, got %s", rawOp)
	}

	op := Op(name)
	declared, known := opFields[op]
	if !known {
		return "", badRequest("unknown op %q", name)
	}

	// Checked ahead of the undeclared-field sweep so the refusal points at load rather
	// than merely naming a field: a caller naming a model wants that model, and running
	// whatever is resident answers a different question.
	if op == OpInfer {
		if _, carriesModel := fields["model"]; carriesModel {
			return "", badRequest("infer takes no model field; load that model first, then infer")
		}
	}

	var undeclared []string
	for field := range fields {
		if field == "op" || slices.Contains(declared, field) {
			continue
		}
		undeclared = append(undeclared, field)
	}
	if len(undeclared) > 0 {
		slices.Sort(undeclared)
		return "", badRequest("%s does not declare the field %q", op, undeclared[0])
	}

	if op == OpLoad {
		id, err := loadTarget(fields)
		if err != nil {
			return "", err
		}
		if id == "" {
			return "", badRequest("load needs an id naming the model to make resident")
		}
	}
	return op, nil
}

// loadTarget reads the model id from a load control block. id is canonical and model is
// accepted as an alias; load is the only op where that field means anything.
func loadTarget(fields map[string]json.RawMessage) (string, error) {
	for _, field := range []string{"id", "model"} {
		raw, ok := fields[field]
		if !ok {
			continue
		}
		var value string
		if err := json.Unmarshal(raw, &value); err != nil {
			return "", badRequest("load %s must be a string, got %s", field, raw)
		}
		if value != "" {
			return value, nil
		}
	}
	return "", nil
}
